nucleus/payara-modules/notification-core/src/main/java/fish/payara/nucleus/notification/NotifierHandler.java

```diff
@@ -0,0 +1,119 @@
+ * Version 2] license."  If you don't indicate a single choice of license, a
+ * recipient has the option to distribute your version of this file under
+ * either the CDDL, the GPL Version 2 or to extend the choice of license to
+ * its licensees as provided above.  However, if you add GPL Version 2 code
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package fish.payara.nucleus.notification;
+
+import static fish.payara.internal.notification.NotifierUtils.getNotifierName;
+import static java.lang.Boolean.valueOf;
+import static java.lang.String.format;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.function.Consumer;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.glassfish.hk2.api.ServiceHandle;
+
+import fish.payara.internal.notification.PayaraConfiguredNotifier;
+import fish.payara.internal.notification.PayaraNotification;
+import fish.payara.internal.notification.PayaraNotifier;
+import fish.payara.internal.notification.PayaraNotifierConfiguration;
+
+public class NotifierHandler implements Runnable, Consumer<PayaraNotification> {
+
+    private static final Logger LOGGER = Logger.getLogger(NotifierHandler.class.getName());
+
+    private final PayaraNotifier notifier;
+    private final String notifierName;
+
+    private final Queue<PayaraNotification> notificationQueue;
+
+    public NotifierHandler(ServiceHandle<PayaraNotifier> notifierHandle) {
+        this(notifierHandle, null);
+    }
+
+    public NotifierHandler(ServiceHandle<PayaraNotifier> notifierHandle, PayaraNotifierConfiguration config) {
+        this.notifier = notifierHandle.getService();
+        this.notifierName = getNotifierName(notifierHandle.getActiveDescriptor());
+        this.notificationQueue = new ConcurrentLinkedQueue<>();
+    }
+
+    protected String getName() {
+        return notifierName;
+    }
+
+    protected void destroy() {
+        notifier.destroy();
+    }
+
+    protected void bootstrap() {
+        notifier.bootstrap();
+    }
+
+    @Override
+    public void accept(PayaraNotification notification) {
+        if (isEnabled() && !this.notificationQueue.offer(notification)) {
+            LOGGER.warning(format("Notifier %s failed to accept the notification \"%s\".", notifierName, notification));
+        }
+    }
+
+    @Override
+    public void run() {
+        final PayaraNotification notification = notificationQueue.peek();
+        try {
+            if (notification != null && isEnabled()) {
+                notifier.handleNotification(notification);
+                notificationQueue.remove();
+            }
+        } catch (Exception ex) {
+            LOGGER.log(Level.WARNING,
+                    format("Notifier %s failed to handle notification \"%s\".", notifierName, notification), ex);
+        }
```

Comment:
It's worth bearing in mind that each thread run loops through every single notifier, so I wasn't sure if trying to flush every single notification every time might have the potential to stall the other notifiers

Reply:
I understand your concern but I can't see how the notifiers actually would affect each other directly other then competing for CPU in general. If we want to enforce fairness beyond OS multi-threading I suggest to look into a single worker thread multi-dispatch where any message enters a single queue and the worker thread pulls from there and dispatches directly to the `PayaraNotifier`s. So instead of a `NotifierHandler` per type of `PayaraNotifier` there is only the single dispatch for all of them.

Reply:
Oh I wasn't looking close enough and assumed from your comment that the notifiers would each be scheduled so that they actually do compete with each other in some form - Now I had a closer look and saw that there is only a single job scheduled that sequentially works the notifiers. So your concern was that working a single notifier in its entirety could delay working the other notifiers later in the sequence.

So what I suggested was more of a round robin. Forget the `NotifierHandler` class exist. Think that in the service you have a list of notifiers and a (generic) queue of messages to deliver. The method called periodically has a loop that runs until the queue is empty. It pulls one message from the queue and forwards it to each notifier that should receive it. This way all notifier will always see the first message before any can see the second which can be said is quite fair. On the downside a notifier that is very slow in its message processing will slow down message processing for all other notifiers. A problem the current loop has as well just in another form. 

Which brings me back to what I thought initially the solution was - that each notifier would have its own queue and its own scheduled task. As this would solve "stalling" messages for other notifiers. Each would in their task method try to send all messages of the queue. Now the notifiers compete for CPU and co but given we want messages going through the system as fast as possible this is a good thing. But we still might delay message processing because of the scheduling. This makes me think: why do we use scheduled tasks when it can be event-reaction that are asynchronously coupled by a queue.

I think what we want is:
* the thread creating a message to be send should not be delayed
* notifiers do not affect each other message delivery
* messages are delivered as promptly as possible
* messages are delivered in order they are handed to the service (for each notifier)

With this in mind I think a queue already decouples a message source from sending the message so the scheduling is just causing unnecessary delay and work. The message processor or processors should just block wait on the queue. A solution that has all the qualities above is a bit more complicated. 
1. Source puts messages in a common queue.
2. a worker block waits on that queue, it takes the message out of the queue and puts it in a queue for each notifier.
3. each notifier specific queue also has a worker task block waiting on the queue that forwards the message to the notifier
4. notifier does its notifier specific thing

Only downside I see here is that this needs n+1 worker threads for n notifier types. But they mostly be waiting.
This can be reduced and made more dynamically scaling by not doing one second level queue for each notifier but a second level queue with work items where a work item is created for each pair of message and target notifier. In that case the sequence is:

1. Source puts messages in a common queue.
2. a worker block waits on that queue, it takes the message out of the first queue, creates a work item each receiving notifier and puts it in the second level queue.
3. another (work stealing) pool of workers is used to process these work items as fast as possible (no scheduling). 

Should there be slow items because of slow notifier implementations the workers in the pool "steal" the work of later items in the queue to some degree. Downside is that order can no longer be guaranteed. Which makes me think that 1 thread per notifier is the best option as usually the number of actually enabled notifiers is low.

Reply:
I've created a thread for each notifier that runs with a 5ms delay but waits on the queue to fill up - I hope this addresses the threading problems!